zebra-chain/src/mmr.rs

```diff
@@ -0,0 +1,190 @@
+
+use crate::{
+    block::{Block, ChainHistoryMmrRootHash},
+    orchard,
+    parameters::{Network, NetworkUpgrade},
+    primitives::zcash_history::{Entry, Tree},
+    sapling,
+};
+
+/// An error describing why a history tree operation failed.
+#[derive(Debug, Error)]
+#[non_exhaustive]
+#[allow(missing_docs)]
+pub enum HistoryTreeError {
+    #[error("error from the underlying library: {inner:?}")]
+    #[non_exhaustive]
+    InnerError { inner: zcash_history::Error },
+
+    #[error("I/O error")]
+    IOError(#[from] io::Error),
+}
+
+/// History tree structure.
+pub struct HistoryTree {
+    network: Network,
+    network_upgrade: NetworkUpgrade,
+    // Merkle mountain range tree.
+    // This is a "runtime" structure used to add / remove nodes, and it's not
+    // persistent.
+    inner: Tree,
+    // The number of nodes in the tree.
+    size: u32,
+    // The peaks of the tree, indexed by their position in the array representation
+    // of the tree. This can be persisted to save the tree.
+    // TODO: use NodeData instead of Entry to save space? Requires re-deriving
+    // the entry metadata from its position.
+    peaks: HashMap<u32, Entry>,
+}
+
+impl HistoryTree {
+    /// Create a new history tree with a single block.
+    pub fn new_from_block(
+        network: Network,
+        block: Arc<Block>,
+        sapling_root: &sapling::tree::Root,
+        _orchard_root: Option<&orchard::tree::Root>,
+    ) -> Result<Self, io::Error> {
+        let height = block
+            .coinbase_height()
+            .expect("block must have coinbase height during contextual verification");
+        let network_upgrade = NetworkUpgrade::current(network, height);
+        // TODO: handle Orchard root
+        let (tree, entry) = Tree::new_from_block(network, block, sapling_root)?;
+        let mut peaks = HashMap::new();
+        peaks.insert(0u32, entry);
+        Ok(HistoryTree {
+            network,
+            network_upgrade,
+            inner: tree,
+            size: 1,
+            peaks,
+        })
+    }
+
+    /// Add block data to the tree.
+    pub fn push(
+        &mut self,
+        block: Arc<Block>,
+        sapling_root: &sapling::tree::Root,
+        _orchard_root: Option<&orchard::tree::Root>,
+    ) -> Result<(), HistoryTreeError> {
+        // TODO: handle orchard root
+        let new_entries = self
+            .inner
+            .append_leaf(block, sapling_root)
+            .map_err(|e| HistoryTreeError::InnerError { inner: e })?;
+        for entry in new_entries {
+            // Not every entry is a peak; those will be trimmed later
+            self.peaks.insert(self.size, entry);
+            self.size += 1;
+        }
+        // TODO: should we really prune every iteration?
+        self.prune()?;
+        // TODO: implement network upgrade logic: drop previous history, start new history
+        Ok(())
+    }
+
+    /// Extend the history tree with the given blocks.
+    pub fn extend<
```

Comment:
API naming: fallible equivalents of standard methods usually start with `try_`
```suggestion
    pub fn try_extend<
```